File: app/src/main/java/com/lifedawn/bestweather/notification/daily/fragment/DailyPushNotificationListFragment.java
package com.lifedawn.bestweather.notification.daily.fragment;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.DialogInterface;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.PopupMenu;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewTreeObserver;
import android.widget.CompoundButton;
import android.widget.RelativeLayout;
import android.widget.Toast;

import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.lifedawn.bestweather.R;
import com.lifedawn.bestweather.commons.classes.RecyclerViewItemDecoration;
import com.lifedawn.bestweather.commons.enums.BundleKey;
import com.lifedawn.bestweather.commons.enums.LocationType;
import com.lifedawn.bestweather.commons.interfaces.OnCheckedSwitchInListListener;
import com.lifedawn.bestweather.commons.interfaces.OnClickedListViewItemListener;
import com.lifedawn.bestweather.commons.interfaces.OnClickedPopupMenuItemListener;
import com.lifedawn.bestweather.databinding.FragmentDailyPushNotificationListBinding;
import com.lifedawn.bestweather.databinding.ViewDailyPushNotificationItemBinding;
import com.lifedawn.bestweather.main.MyApplication;
import com.lifedawn.bestweather.notification.daily.DailyNotificationHelper;
import com.lifedawn.bestweather.notification.daily.DailyPushNotificationType;
import com.lifedawn.bestweather.notification.daily.viewmodel.DailyNotificationViewModel;
import com.lifedawn.bestweather.room.callback.DbQueryCallback;
import com.lifedawn.bestweather.room.dto.DailyPushNotificationDto;
import com.lifedawn.bestweather.room.repository.DailyPushNotificationRepository;

import org.jetbrains.annotations.NotNull;

import java.time.LocalTime;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;


public class DailyPushNotificationListFragment extends Fragment {
	private FragmentDailyPushNotificationListBinding binding;
	private NotificationListAdapter adapter;
	private final DateTimeFormatter hoursFormatter = DateTimeFormatter.ofPattern("a h:mm");
	private DailyPushNotificationRepository repository;
	private DailyNotificationHelper dailyNotificationHelper;
	private DailyNotificationViewModel dailyNotificationViewModel;
	private final Comparator<DailyPushNotificationDto> sortComparator = new Comparator<DailyPushNotificationDto>() {
		@Override
		public int compare(DailyPushNotificationDto o1, DailyPushNotificationDto o2) {
			return LocalTime.parse(o1.getAlarmClock()).compareTo(LocalTime.parse(o2.getAlarmClock()));
		}
	};

	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		repository = DailyPushNotificationRepository.getINSTANCE();
		dailyNotificationHelper = new DailyNotificationHelper(requireContext().getApplicationContext());
		dailyNotificationViewModel = new ViewModelProvider(this).get(DailyNotificationViewModel.class);
	}

	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container,
	                         Bundle savedInstanceState) {
		binding = FragmentDailyPushNotificationListBinding.inflate(inflater);

		RelativeLayout.LayoutParams layoutParams = (RelativeLayout.LayoutParams) binding.toolbar.getRoot().getLayoutParams();
		layoutParams.topMargin = MyApplication.getStatusBarHeight();
		binding.toolbar.getRoot().setLayoutParams(layoutParams);

		binding.progressResultView.setContentView(binding.notificationList);
		binding.toolbar.fragmentTitle.setText(R.string.daily_notification);
		binding.toolbar.backBtn.setOnClickListener(new View.OnClickListener() {
			@Override
			public void onClick(View v) {
				getParentFragmentManager().popBackStack();
			}
		});

		binding.notificationList.setLayoutManager(new LinearLayoutManager(getContext(), RecyclerView.VERTICAL, false));

		return binding.getRoot();
	}

	@Override
	public void onViewCreated(@NonNull @NotNull View view, @Nullable @org.jetbrains.annotations.Nullable Bundle savedInstanceState) {
		super.onViewCreated(view, savedInstanceState);


		binding.getRoot().getViewTreeObserver().addOnGlobalLayoutListener(new ViewTreeObserver.OnGlobalLayoutListener() {
			@Override
			public void onGlobalLayout() {
				binding.getRoot().getViewTreeObserver().removeOnGlobalLayoutListener(this);

				binding.notificationList.addItemDecoration(new RecyclerViewItemDecoration(requireContext().getApplicationContext(), true,
						binding.getRoot().getHeight() - binding.addBtn.getTop() - binding.addBtn.getHeight() / 2));
			}
		});


		adapter = new NotificationListAdapter(requireContext().getApplicationContext(), new OnClickedListViewItemListener<DailyPushNotificationDto>() {
			@Override
			public void onClickedItem(DailyPushNotificationDto e) {
				DailyNotificationSettingsFragment settingsFragment = new DailyNotificationSettingsFragment();
				String tag = DailyNotificationSettingsFragment.class.getName();

				Bundle bundle = new Bundle();
				bundle.putSerializable("dto", e);
				bundle.putBoolean(BundleKey.NewSession.name(), false);
				settingsFragment.setArguments(bundle);

				getParentFragmentManager().beginTransaction().hide(DailyPushNotificationListFragment.this).add(R.id.fragment_container,
						settingsFragment, tag).addToBackStack(tag).setPrimaryNavigationFragment(settingsFragment).commitAllowingStateLoss();
			}
		}, new OnCheckedSwitchInListListener<DailyPushNotificationDto>() {
			@Override
			public void onCheckedSwitch(DailyPushNotificationDto dailyPushNotificationDto, boolean isChecked) {
				dailyPushNotificationDto.setEnabled(isChecked);
				repository.update(dailyPushNotificationDto, null);
				dailyNotificationHelper.modifyPushNotification(dailyPushNotificationDto);

				String text = LocalTime.parse(dailyPushNotificationDto.getAlarmClock()).format(hoursFormatter)
						+ ", ";

				if (isChecked) {
					text += getString(R.string.registeredDailyNotification);
				} else {
					text += getString(R.string.unregisteredDailyNotification);
				}
				Toast.makeText(getContext(), text, Toast.LENGTH_SHORT).show();
			}
		}, new OnClickedPopupMenuItemListener<DailyPushNotificationDto>() {
			@Override
			public void onClickedItem(DailyPushNotificationDto e, int position) {
				switch (position) {
					case 0:
						new MaterialAlertDialogBuilder(requireActivity())
								.setTitle(R.string.removeNotification)
								.setMessage(R.string.will_you_delete_the_notification)
								.setPositiveButton(R.string.remove, new DialogInterface.OnClickListener() {
									@Override
									public void onClick(DialogInterface dialog, int which) {
										dailyNotificationHelper.disablePushNotification(e);
										repository.delete(e, null);
										dialog.dismiss();
									}
								}).setNegativeButton(R.string.cancel, new DialogInterface.OnClickListener() {
									@Override
									public void onClick(DialogInterface dialog, int which) {
										dialog.dismiss();
									}
								}).create().show();
						break;
				}
			}
		});
		adapter.registerAdapterDataObserver(new RecyclerView.AdapterDataObserver() {
			@Override
			public void onChanged() {
				super.onChanged();
				if (adapter.getItemCount() > 0) {
					binding.progressResultView.onSuccessful();
				} else {
					binding.progressResultView.onFailed(getString(R.string.empty_daily_notifications));
				}
			}
		});
		binding.notificationList.setAdapter(adapter);

		binding.addBtn.setOnClickListener(new View.OnClickListener() {
			@Override
			public void onClick(View v) {
				DailyNotificationSettingsFragment settingsFragment = new DailyNotificationSettingsFragment();
				String tag = DailyNotificationSettingsFragment.class.getName();

				Bundle bundle = new Bundle();
				bundle.putBoolean(BundleKey.NewSession.name(), true);
				settingsFragment.setArguments(bundle);

				getParentFragmentManager().beginTransaction().hide(DailyPushNotificationListFragment.this).add(R.id.fragment_container,
						settingsFragment, tag).addToBackStack(tag).commit();
			}
		});

		dailyNotificationViewModel.listLiveData.observe(getViewLifecycleOwner(), new Observer<List<DailyPushNotificationDto>>() {
			@Override
			public void onChanged(List<DailyPushNotificationDto> result) {
				Collections.sort(result, sortComparator);
				adapter.setNotificationList(result);
				adapter.notifyDataSetChanged();
			}
		});
	}

	@Override
	public void onDestroyView() {
		super.onDestroyView();
		binding = null;
	}


	@Override
	public void onHiddenChanged(boolean hidden) {
		super.onHiddenChanged(hidden);
	}


	public static class NotificationListAdapter extends RecyclerView.Adapter<NotificationListAdapter.ViewHolder> {
		private List<DailyPushNotificationDto> notificationList = new ArrayList<>();
		private Context context;
		private OnClickedListViewItemListener<DailyPushNotificationDto> onClickedItemListener;
		private OnCheckedSwitchInListListener<DailyPushNotificationDto> onCheckedSwitchInListListener;
		private OnClickedPopupMenuItemListener<DailyPushNotificationDto> onClickedPopupMenuItemListener;
		private final DateTimeFormatter hoursFormatter = DateTimeFormatter.ofPattern("a hh:mm");

		public NotificationListAdapter(Context context, OnClickedListViewItemListener<DailyPushNotificationDto> onClickedItemListener,
		                               OnCheckedSwitchInListListener<DailyPushNotificationDto> onCheckedSwitchInListListener,
		                               OnClickedPopupMenuItemListener<DailyPushNotificationDto> onClickedPopupMenuItemListener) {
			this.onClickedItemListener = onClickedItemListener;
			this.context = context;
			this.onCheckedSwitchInListListener = onCheckedSwitchInListListener;
			this.onClickedPopupMenuItemListener = onClickedPopupMenuItemListener;
		}

		public void setNotificationList(List<DailyPushNotificationDto> notificationList) {
			this.notificationList = notificationList;
		}

		@NonNull
		@NotNull
		@Override
		public ViewHolder onCreateViewHolder(@NonNull @NotNull ViewGroup parent, int viewType) {
			return new ViewHolder(LayoutInflater.from(context).inflate(R.layout.view_daily_push_notification_item, parent, false));
		}

		@Override
		public void onBindViewHolder(@NonNull @NotNull DailyPushNotificationListFragment.NotificationListAdapter.ViewHolder holder, int position) {
			holder.onBind();
		}

		@Override
		public int getItemCount() {
			return notificationList.size();
		}

		private class ViewHolder extends RecyclerView.ViewHolder {
			private ViewDailyPushNotificationItemBinding binding;
			private boolean init = true;

			public ViewHolder(@NonNull @NotNull View itemView) {
				super(itemView);
				binding = ViewDailyPushNotificationItemBinding.bind(itemView);

				binding.notiSwitch.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {

					@Override
					public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
						if (init) {
							return;
						}
						onCheckedSwitchInListListener.onCheckedSwitch(notificationList.get(getBindingAdapterPosition()), isChecked);
					}

				});

				binding.getRoot().setOnClickListener(new View.OnClickListener() {
					@Override
					public void onClick(View v) {
						onClickedItemListener.onClickedItem(notificationList.get(getBindingAdapterPosition()));
					}
				});


				binding.control.setOnClickListener(new View.OnClickListener() {
					@Override
					public void onClick(View v) {
						PopupMenu popupMenu = new PopupMenu(context, binding.control, Gravity.BOTTOM);

						popupMenu.getMenuInflater().inflate(R.menu.menu_of_daily_notification_item, popupMenu.getMenu());
						popupMenu.setOnMenuItemClickListener(new PopupMenu.OnMenuItemClickListener() {
							@SuppressLint("NonConstantResourceId")
							@Override
							public boolean onMenuItemClick(MenuItem menuItem) {
								switch (menuItem.getItemId()) {
									default:
										onClickedPopupMenuItemListener.onClickedItem(notificationList.get(getBindingAdapterPosition()), getBindingAdapterPosition());
								}
								return true;
							}
						});

						popupMenu.show();
					}
				});
			}

			public void onBind() {
				init = true;
				final int pos = getBindingAdapterPosition();
				DailyPushNotificationDto dto = notificationList.get(pos);

				binding.hours.setText(LocalTime.parse(dto.getAlarmClock()).format(hoursFormatter));
				binding.notiSwitch.setChecked(dto.isEnabled());

				binding.notificationType.setText(DailyPushNotificationType.getNotificationName(dto.getNotificationType(), context));

				String addressName = dto.getLocationType() == LocationType.SelectedAddress ? dto.getAddressName() :
						context.getString(R.string.current_location);
				binding.location.setText(addressName);
				init = false;
			}

		}
	}
}

File: app/src/main/java/com/lifedawn/bestweather/ui/notification/daily/fragment/DailyPushNotificationListFragment.kt
package com.lifedawn.bestweather.ui.notification.daily.fragment

import android.os.Bundle
import android.view.*
import android.widget.CompoundButton
import android.widget.RelativeLayout
import android.widget.Toast
import androidx.appcompat.widget.PopupMenu
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.AdapterDataObserver
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.lifedawn.bestweather.R
import com.lifedawn.bestweather.commons.classes.RecyclerViewItemDecoration
import com.lifedawn.bestweather.commons.constants.BundleKey
import com.lifedawn.bestweather.commons.constants.LocationType
import com.lifedawn.bestweather.commons.interfaces.OnCheckedSwitchInListListener
import com.lifedawn.bestweather.commons.interfaces.OnClickedListViewItemListener
import com.lifedawn.bestweather.commons.interfaces.OnClickedPopupMenuItemListener
import com.lifedawn.bestweather.data.MyApplication
import com.lifedawn.bestweather.data.local.room.dto.DailyPushNotificationDto
import com.lifedawn.bestweather.data.local.room.repository.DailyPushNotificationRepository
import com.lifedawn.bestweather.databinding.FragmentDailyPushNotificationListBinding
import com.lifedawn.bestweather.databinding.ViewDailyPushNotificationItemBinding
import com.lifedawn.bestweather.ui.notification.daily.DailyNotificationHelper
import com.lifedawn.bestweather.ui.notification.daily.DailyPushNotificationType
import com.lifedawn.bestweather.ui.notification.daily.viewmodel.DailyNotificationViewModel
import java.time.LocalTime
import java.time.format.DateTimeFormatter
import java.util.*

class DailyPushNotificationListFragment : Fragment() {
    private var binding: FragmentDailyPushNotificationListBinding? = null
    private var adapter: NotificationListAdapter? = null
    private val hoursFormatter = DateTimeFormatter.ofPattern("a h:mm")
    private var repository: DailyPushNotificationRepository? = null
    private var dailyNotificationHelper: DailyNotificationHelper? = null
    private var dailyNotificationViewModel: DailyNotificationViewModel? = null
    private val sortComparator =
        Comparator<DailyPushNotificationDto> { o1, o2 -> LocalTime.parse(o1.alarmClock).compareTo(LocalTime.parse(o2.alarmClock)) }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        repository = DailyPushNotificationRepository.getINSTANCE()
        dailyNotificationHelper = DailyNotificationHelper(requireContext().applicationContext)
        dailyNotificationViewModel = ViewModelProvider(this).get(DailyNotificationViewModel::class.java)
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        binding = FragmentDailyPushNotificationListBinding.inflate(inflater, container, false)
        val layoutParams = binding!!.toolbar.root.layoutParams as RelativeLayout.LayoutParams
        layoutParams.topMargin = MyApplication.getStatusBarHeight()
        binding!!.toolbar.root.layoutParams = layoutParams
        binding!!.progressResultView.setContentView(binding!!.notificationList)
        binding!!.toolbar.fragmentTitle.setText(R.string.daily_notification)
        binding!!.toolbar.backBtn.setOnClickListener { parentFragmentManager.popBackStack() }
        binding!!.notificationList.layoutManager = LinearLayoutManager(context, RecyclerView.VERTICAL, false)
        return binding!!.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding!!.root.viewTreeObserver.addOnGlobalLayoutListener(object : ViewTreeObserver.OnGlobalLayoutListener {
            override fun onGlobalLayout() {
                binding!!.root.viewTreeObserver.removeOnGlobalLayoutListener(this)
                binding!!.notificationList.addItemDecoration(
                    RecyclerViewItemDecoration(
                        requireContext().applicationContext, true,
                        binding!!.root.height - binding!!.addBtn.top - binding!!.addBtn.height / 2
                    )
                )
            }
        })
        adapter = NotificationListAdapter(object : OnClickedListViewItemListener<DailyPushNotificationDto?> {
            override fun onClickedItem(e: DailyPushNotificationDto) {
                val settingsFragment = DailyNotificationSettingsFragment()
                val tag = DailyNotificationSettingsFragment::class.java.name
                val bundle = Bundle()
                bundle.putSerializable("dto", e)
                bundle.putBoolean(BundleKey.NewSession.name, false)
                settingsFragment.arguments = bundle
                parentFragmentManager.beginTransaction().hide(this@DailyPushNotificationListFragment).add(
                    R.id.fragment_container,
                    settingsFragment, tag
                ).addToBackStack(tag).setPrimaryNavigationFragment(settingsFragment).commitAllowingStateLoss()
            }
        }, object : OnCheckedSwitchInListListener<DailyPushNotificationDto?> {
            override fun onCheckedSwitch(dailyPushNotificationDto: DailyPushNotificationDto, isChecked: Boolean) {
                dailyPushNotificationDto.isEnabled = isChecked
                repository!!.update(dailyPushNotificationDto, null)
                dailyNotificationHelper!!.modifyPushNotification(dailyPushNotificationDto)
                var text = (LocalTime.parse(dailyPushNotificationDto.alarmClock).format(hoursFormatter)
                        + ", ")
                text += if (isChecked) {
                    getString(R.string.registeredDailyNotification)
                } else {
                    getString(R.string.unregisteredDailyNotification)
                }
                Toast.makeText(context, text, Toast.LENGTH_SHORT).show()
            }
        }, object : OnClickedPopupMenuItemListener<DailyPushNotificationDto?> {
            override fun onClickedItem(e: DailyPushNotificationDto, position: Int) {
                when (position) {
                    0 -> MaterialAlertDialogBuilder(requireActivity())
                        .setTitle(R.string.removeNotification)
                        .setMessage(R.string.will_you_delete_the_notification)
                        .setPositiveButton(R.string.remove) { dialog, which ->
                            dailyNotificationHelper!!.disablePushNotification(e)
                            repository!!.delete(e, null)
                            dialog.dismiss()
                        }
                        .setNegativeButton(R.string.cancel) { dialog, which -> dialog.dismiss() }.create().show()
                }
            }
        })
        adapter!!.registerAdapterDataObserver(object : AdapterDataObserver() {
            override fun onChanged() {
                super.onChanged()
                if (adapter!!.itemCount > 0) {
                    binding!!.progressResultView.onSuccessful()
                } else {
                    binding!!.progressResultView.onFailed(getString(R.string.empty_daily_notifications))
                }
            }
        })
        binding!!.notificationList.adapter = adapter
        binding!!.addBtn.setOnClickListener {
            val settingsFragment = DailyNotificationSettingsFragment()
            val tag = DailyNotificationSettingsFragment::class.java.name
            val bundle = Bundle()
            bundle.putBoolean(BundleKey.NewSession.name, true)
            settingsFragment.arguments = bundle
            parentFragmentManager.beginTransaction().hide(this@DailyPushNotificationListFragment).add(
                R.id.fragment_container,
                settingsFragment, tag
            ).addToBackStack(tag).commit()
        }
        dailyNotificationViewModel!!.listLiveData.observe(viewLifecycleOwner) { result ->
            Collections.sort(result, sortComparator)
            adapter!!.setNotificationList(result)
            adapter!!.notifyDataSetChanged()
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        binding = null
    }

    override fun onHiddenChanged(hidden: Boolean) {
        super.onHiddenChanged(hidden)
    }

    class NotificationListAdapter(
        private val onClickedItemListener: OnClickedListViewItemListener<DailyPushNotificationDto>,
        private val onCheckedSwitchInListListener: OnCheckedSwitchInListListener<DailyPushNotificationDto>,
        private val onClickedPopupMenuItemListener: OnClickedPopupMenuItemListener<DailyPushNotificationDto>
    ) : RecyclerView.Adapter<NotificationListAdapter.ViewHolder>() {
        private var notificationList: List<DailyPushNotificationDto> = ArrayList()
        private val hoursFormatter = DateTimeFormatter.ofPattern("a hh:mm")
        fun setNotificationList(notificationList: List<DailyPushNotificationDto>) {
            this.notificationList = notificationList
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
            return ViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.view_daily_push_notification_item, parent, false))
        }

        override fun onBindViewHolder(holder: ViewHolder, position: Int) {
            holder.onBind()
        }

        override fun getItemCount(): Int {
            return notificationList.size
        }

        inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
            private val binding: ViewDailyPushNotificationItemBinding
            private var init = true

            init {
                binding = ViewDailyPushNotificationItemBinding.bind(itemView)
                binding.notiSwitch.setOnCheckedChangeListener(CompoundButton.OnCheckedChangeListener { buttonView, isChecked ->
                    if (init) {
                        return@OnCheckedChangeListener
                    }
                    onCheckedSwitchInListListener.onCheckedSwitch(notificationList[bindingAdapterPosition], isChecked)
                })
                binding.root.setOnClickListener { v: View? -> onClickedItemListener.onClickedItem(notificationList[bindingAdapterPosition]) }
                binding.control.setOnClickListener { v: View? ->
                    val popupMenu = PopupMenu(binding.root.context, binding.control, Gravity.BOTTOM)
                    popupMenu.menuInflater.inflate(R.menu.menu_of_daily_notification_item, popupMenu.menu)
                    popupMenu.setOnMenuItemClickListener(PopupMenu.OnMenuItemClickListener { menuItem ->
                        when (menuItem.itemId) {
                            else -> onClickedPopupMenuItemListener.onClickedItem(
                                notificationList[bindingAdapterPosition],
                                bindingAdapterPosition
                            )
                        }
                        return@setOnClickListener true
                    })
                    popupMenu.show()
                }
            }

            fun onBind() {
                init = true
                val pos = bindingAdapterPosition
                val dto = notificationList[pos]
                binding.hours.text = LocalTime.parse(dto.alarmClock).format(hoursFormatter)
                binding.notiSwitch.isChecked = dto.isEnabled
                binding.notificationType.setText(
                    DailyPushNotificationType.Companion.getNotificationName(
                        dto.notificationType,
                        binding.root.context.applicationContext
                    )
                )
                val addressName =
                    if (dto.locationType === LocationType.SelectedAddress) dto.addressName else binding.root.context.getString(
                        R.string.current_location
                    )
                binding.location.text = addressName
                init = false
            }
        }
    }
}